File: android/src/main/java/com/adyenreactnativesdk/configuration/ThreeDSConfigurationParser.kt
/*
 * Copyright (c) 2023 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 */

package com.adyenreactnativesdk.configuration

import com.adyen.checkout.authentication.authentication
import com.adyen.checkout.core.components.CheckoutConfiguration
import com.facebook.react.bridge.ReadableMap

class ThreeDSConfigurationParser(
  config: ReadableMap,
) {
  companion object {
    const val TAG = "ThreeDSConfigurationParser"
    const val ROOT_KEY = "threeDS2"
    const val REQUESTOR_APP_URL_KEY = "requestorAppUrl"
  }

  private var config: ReadableMap

  init {
    if (config.hasKey(ROOT_KEY)) {
      this.config = config.getMap(ROOT_KEY)!!
    } else {
      this.config = config
    }
  }

  internal val requestorAppUrl: String?
    get() =
      if (config.hasKey(REQUESTOR_APP_URL_KEY)) {
        config.getString(REQUESTOR_APP_URL_KEY)
      } else {
        null
      }

  fun applyConfiguration(configuration: CheckoutConfiguration) {
    configuration.authentication(threeDSRequestorAppURL = requestorAppUrl)
  }
}

File: ios/Configuration/ThreeDSConfigurationParser.swift
//
// Copyright (c) 2023 Adyen N.V.
//
// This file is open source and available under the MIT license. See the LICENSE file for more info.
//

import Adyen
import Foundation

public struct ThreeDS2ConfigurationParser {

    private var dict: NSDictionary

    public init(configuration: NSDictionary) {
        if let configurationNode = configuration[ThreeDSKey.rootKey] as? NSDictionary {
            self.dict = configurationNode
        } else {
            self.dict = configuration
        }
    }

    var requestorAppUrl: String? {
        dict[ThreeDSKey.requestorAppUrl] as? String
    }

    public var configuration: AuthenticationConfiguration {
        var configuration = AuthenticationConfiguration()
        if let requestorAppUrl, let url = URL(string: requestorAppUrl) {
            configuration = configuration.requestorAppURL(url)
        }
        return configuration
    }
}
